### PR TITLE
fix: Qute template validation breaks when using Gradle

### LINKED ISSUE #1224: Qute template validation breaks when using Gradle
There are no issues with the templates and everything works correctly but I am getting errors with the Qute validation.

image

image

The templates are located here in the build directly and they do work as expected.

image

### PR DESCRIPTION
**Pull request by angelozerr**
fix: Qute template validation breaks when using Gradle

Fixes #1224

### CODE REVIEW

**Comment by fbricon** [comment]
Doesn't work. When I put a breakpoint on the [calling code](https://github.com/redhat-developer/intellij-quarkus/blob/b3252cc7148d62b814c4de4471e56c93126e4a4d/src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java#L53), to display the value of templateDir, I get these logged, when opening the java class declaring the template:

```
file:/Users/fbricon/Dev/souk/grute/.idea/src/main/resources/templates/
file:/Users/fbricon/Dev/souk/grute/.idea/src/main/resources/templates/
file:/Users/fbricon/Dev/souk/grute/src/main/resources/templates
file:/Users/fbricon/Dev/souk/grute/.idea/src/main/resources/templates/
```

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/LSPIJUtils.java [hunk 1]
```diff
@@ -116,40 +114,90 @@ public static URI toUri(File file) {
         }
     }
 
-    public static URI toUri(PsiFile file) {
-        return toUri(file.getVirtualFile());
+
+    public static @NotNull URI toUri(@NotNull Module module) {
+        VirtualFile[] roots = ModuleRootManager.getInstance(module).getContentRoots();
+        if (roots.length > 0) {
+            return toUri(roots[0]);
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/LSPIJUtils.java:121]: this is not enough. The generated-sources content root shows up 1st in a myproject.main module

File: src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java [hunk 2]
```diff
@@ -60,6 +61,10 @@ public static ProjectInfo getProjectInfo(Module javaProject) {
 	}
 
 	private static String getTemplateBaseDir(Module javaProject) {
+		List<VirtualFile> resourcesDirs = ModuleRootManager.getInstance(javaProject).getSourceRoots(JavaResourceRootType.RESOURCE);
+		if (!resourcesDirs.isEmpty()) {
+			return LSPIJUtils.toUri(resourcesDirs.get(0)).resolve("templates").toASCIIString();
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java]: fixed

**Comment by fbricon** [comment]
There are at least 2 problems still, related to missing templates:

- the create template codelens generates a template under build/classes/java/quarkus-generated-sources/ 
- On ctrl+click on a missing template field, it generates a .html.html file:

Screenshot 2023-10-30 at 17 15 54

**Comment by fbricon** [comment]
After moving the file created in the generated source folder to the proper src/main/resources folder, template completion finds stale entries :
Screenshot 2023-10-30 at 17 28 40

File: src/main/java/com/redhat/devtools/intellij/qute/psi/internal/java/AbstractQuteTemplateLinkCollector.java [hunk 3]
```diff
@@ -37,228 +38,250 @@
...
+     * @param module           the module.
+     * @return the relative template file path from the module.
+     */
+    private @NotNull String getRelativePath(@NotNull String templateFilePath, @Nullable VirtualFile templateFile, @NotNull Module module) {
+        if (templateFile != null) {
+            // The template file exists
+            // Compute the relative path from the module root path:
+            // - src/main/resources/templates/page.html for Maven project
+            // - resources/templates/page.html for Gradle project
+            for (VirtualFile root : ModuleRootManager.getInstance(module).getContentRoots()) {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/qute/psi/internal/java/AbstractQuteTemplateLinkCollector.java:244]: You'll likely need to use this trick
https://github.com/redhat-developer/intellij-quarkus/pull/1240/files#diff-75cbc8b0d9b6203597c60ac5fb2193eccbd1308d712e1a880c67d078b21d7889R213

to keep non-generated source folders first

File: projects/qute/projects/gradle/qute-quickstart/gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.1.1-bin.zip
```
Comment (fbricon) [inline projects/qute/projects/gradle/qute-quickstart/gradle/wrapper/gradle-wrapper.properties:3]: use 8.4

File: src/main/java/com/redhat/devtools/intellij/qute/psi/QuteSupportForTemplate.java [hunk 5]
```diff
@@ -326,10 +326,8 @@ public static Module getJavaProjectFromTemplateFile(String templateFileUri, IPsi
 			// The uri belong to an IntelliJ project
 			if (ModuleType.get(module) != JavaModuleType.getModuleType()) {
 				// The uri doesn't belong to a IntelliJ project
-				return null;
+				//return null;
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/qute/psi/QuteSupportForTemplate.java:329]: might as well remove the whole if block

File: src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java [hunk 6]
```diff
@@ -32,104 +33,165 @@
...
+                var templatesDir = dir.findChild(TEMPLATES_FOLDER_NAME);
+                if (templatesDir != null && templatesDir.exists()) {
+                    return dir;
+                }
+            }
+            // 2) no resources directories contains the 'templates' folder,returns the first.
+            return resourcesDirs.get(0);
+        }
+        // Corner usecase, the module doesn't configure 'Resources folder', use the first content roots
+        VirtualFile[] roots = ModuleRootManager.getInstance(javaProject).getContentRoots();
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java:120]: Arrays.sort(roots, Comparator.comparingInt(r -> r.getPath().length()));

**Review by fbricon — CHANGES_REQUESTED** [review]

File: src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java [hunk 7]
```diff
@@ -1,135 +1,196 @@
...
+
+    /**
+     * Returns the best 'resources' directory for the given Java module project and null otherwise.
+     *
+     * @param javaProject the Java module project.
+     *
+     * @return the best resources dir for the given Java module project.
+     */
+    public static @Nullable VirtualFile findBestResourcesDir(@NotNull Module javaProject) {
+        List<VirtualFile> resourcesDirs = ModuleRootManager.getInstance(javaProject).getSourceRoots(JavaResourceRootType.RESOURCE);
```
[the remark this answers is not in the document]
Comment (fbricon) [reply src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java:104]: mm actually, even after removing the resources folder in quarkus-generated-sources, the template is still created there

File: src/main/java/com/redhat/devtools/intellij/qute/psi/internal/java/AbstractQuteTemplateLinkCollector.java [hunk 8]
```diff
@@ -21,244 +21,262 @@
...
+            if (templateFile != null && templateFile.exists()) {
+                return templateFile;
+            }
+        }
+        return templatesDir.findFileByRelativePath(templateUri + PREFERRED_SUFFIX);
+    }
+
+    protected static String getVirtualFileUrl(Module project, String templateFilePath) {
+        try {
+            for (VirtualFile root : ModuleRootManager.getInstance(project).getContentRoots()) {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/qute/psi/internal/java/AbstractQuteTemplateLinkCollector.java:247]: this is where the template URI picks up the generated sources folder, when a template is missing. Needs to be sorted

File: projects/qute/projects/gradle/qute-quickstart/gradle/wrapper/gradle-wrapper.properties (hunk 4 above)
Comment (angelozerr) [reply projects/qute/projects/gradle/qute-quickstart/gradle/wrapper/gradle-wrapper.properties]: fixed

File: src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java (hunk 6 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java]: fixed

File: src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/qute/psi/utils/PsiQuteProjectUtils.java:105]: fixed

File: src/main/java/com/redhat/devtools/intellij/qute/psi/internal/java/AbstractQuteTemplateLinkCollector.java (hunk 8 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/qute/psi/internal/java/AbstractQuteTemplateLinkCollector.java]: fixed

File: src/main/java/com/redhat/devtools/intellij/qute/psi/internal/java/AbstractQuteTemplateLinkCollector.java (hunk 3 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/qute/psi/internal/java/AbstractQuteTemplateLinkCollector.java]: fixed

**Comment by fbricon** [comment]
@angelozerr thanks!

**Pull request merged**